### PR TITLE
[narwhal] introduce primary/worker container structs

### PR DESCRIPTION
**Pull request by akichidis**
This PR is replacing the current way of spawning primary/worker nodes to allow us manager them via dedicated handlers instead of just ending up with a bunch of `JoinHandles` instead. More specifically:
* The `PrimaryNode` , `WorkerNode` and `WorkerNodes` helper structs are introduced
* The above helper structs allow us to easily start/stop a node via simple method calls
* Is now giving us the ability to individually shutdown a primary/worker node without having the need to propagate a `/reconfigure shutdown` signal from the primary to the worker 
* It removes the need to send a `shutdown` signal to a node via a network admin interface, and instead it utilises the introduced broadcast channels to directly send those signals , making the whole process more robust.

### CODE REVIEW

File: narwhal/node/src/worker_node.rs [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+        if let Some(registry) = registry {
+            self.registry = Some((self.registry_service.add(registry.clone()), registry));
+        } else {
+            self.registry = None
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct WorkerNode {
```
Comment (akichidis) [inline narwhal/node/src/worker_node.rs:149]: The `WorkerNode` (similarly to the `PrimaryNode`) structs are wrapper/shell structs around the "Inner" ones to allow us modify the internals using a sharable across threads struct and not restricting us from having to use an exclusive `&mut` object in our code.

File: narwhal/primary/src/state_handler.rs [hunk 2]
```diff
@@ -113,42 +85,15 @@ impl StateHandler {
...
-                    // TODO: this request will be removed https://mysten.atlassian.net/browse/SUI-984
-                    let join_all = futures::future::try_join_all(notify_handlers);
-                    join_all.await.expect("Error while sending reconfiguration message to the workers");
-
-                    warn!("Successfully broadcasted reconfigure message to workers");
-
-                    // Exit only when we are sure that all the other tasks received
-                    // the shutdown message.
-
-                    // shutdown network as well
```
Comment (laura-makdah) [inline narwhal/primary/src/state_handler.rs:139]: 🎉

File: narwhal/node/src/worker_node.rs (hunk 1 above)
Comment (laura-makdah) [reply narwhal/node/src/worker_node.rs:149]: In what way do we intend to modify the internals?

**Review by mwtian — APPROVED** [review]
Looks good at a high level. I wonder if the structures can be simplified but we can iterate on that. Added a fix to `basic_cluster_setup` test. It seems in `test_shutdown_bug`, where only one authority is started then stopped, the shutdown cannot finish. Will take another look tomorrow.

Approve to unblock.

File: narwhal/node/src/worker_node.rs [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+    }
+
+    pub async fn wait(&self) {
+        let mut guard = self.internal.write().await;
+        guard.wait().await
+    }
+}
+
+pub struct WorkerNodes {
+    workers: DashMap<WorkerId, WorkerNode>,
```
[the remark this answers is not in the document]
Comment (mwtian) [reply narwhal/node/src/worker_node.rs]: Updated.

**Pull request merged**